pkg/provider/vault/config.go

```diff
@@ -0,0 +1,406 @@
+	AgentImagePullPolicy          corev1.PullPolicy
+	AgentEnvVariables             string
+	ServiceAccountTokenVolumeName string
+	TokenAuthMount                string
+	VaultNamespace                string
+	VaultServiceAccount           string
+	Token                         string
+	IgnoreMissingSecrets          string
+	Passthrough                   string
+	LogLevel                      string
+	FromPath                      string
+}
+
+func ParseConfig(obj metav1.Object, ar *model.AdmissionReview) (Config, error) {
+	SetDefaults()
+
+	config := Config{
+		ObjectNamespace: ar.Namespace,
+	}
+
+	annotations := obj.GetAnnotations()
+
+	if val, ok := annotations[common.VaultAddrAnnotation]; ok {
+		config.Addr = val
+	} else {
+		config.Addr = viper.GetString("vault_addr")
+	}
+
+	if val, ok := annotations[common.VaultRoleAnnotation]; ok {
+		config.Role = val
+	} else {
+		if val := viper.GetString("vault_role"); val != "" {
+			config.Role = val
+		} else {
+			switch p := obj.(type) {
+			case *corev1.Pod:
+				config.Role = p.Spec.ServiceAccountName
+			default:
+				config.Role = "default"
+			}
+		}
+	}
+
+	if val, ok := annotations[common.VaultAuthMethodAnnotation]; ok {
+		config.AuthMethod = val
+	} else {
+		config.AuthMethod = viper.GetString("vault_auth_method")
+	}
+
+	if val, ok := annotations[common.VaultPathAnnotation]; ok {
+		config.Path = val
+	} else {
+		config.Path = viper.GetString("vault_path")
+	}
+
+	// TODO: Check for flag to verify we want to use namespace-local SAs instead of the webhook namespaces SA
+	if val, ok := annotations[common.VaultServiceaccountAnnotation]; ok {
+		config.VaultServiceAccount = val
+	} else {
+		config.VaultServiceAccount = viper.GetString("vault_serviceaccount")
+	}
+
+	if val, ok := annotations[common.VaultSkipVerifyAnnotation]; ok {
+		config.SkipVerify, _ = strconv.ParseBool(val)
+	} else {
+		config.SkipVerify = viper.GetBool("vault_skip_verify")
+	}
+
+	if val, ok := annotations[common.VaultTLSSecretAnnotation]; ok {
+		config.TLSSecret = val
+	} else {
+		config.TLSSecret = viper.GetString("vault_tls_secret")
+	}
+
+	if val, ok := annotations[common.VaultClientTimeoutAnnotation]; ok {
+		config.ClientTimeout, _ = time.ParseDuration(val)
+	} else {
+		config.ClientTimeout, _ = time.ParseDuration(viper.GetString("vault_client_timeout"))
+	}
+
+	if val, ok := annotations[common.VaultAgentAnnotation]; ok {
+		config.UseAgent, _ = strconv.ParseBool(val)
+	} else {
+		config.UseAgent, _ = strconv.ParseBool(viper.GetString("vault_agent"))
+	}
+
+	if val, ok := annotations[common.VaultConsulTemplateConfigmapAnnotation]; ok {
+		config.CtConfigMap = val
+	} else {
+		config.CtConfigMap = ""
+	}
+
+	if val, ok := annotations[common.ServiceAccountTokenVolumeNameAnnotation]; ok {
+		config.ServiceAccountTokenVolumeName = val
+	} else if viper.GetString("SERVICE_ACCOUNT_TOKEN_VOLUME_NAME") != "" {
+		config.ServiceAccountTokenVolumeName = viper.GetString("SERVICE_ACCOUNT_TOKEN_VOLUME_NAME")
```

Comment:
env-vars should also be part of common package (env-specific consts can be suffixed with `...Env`)